crates/oxc_transformer/src/typescript/enum.rs

```diff
@@ -227,4 +280,262 @@ impl<'a> TypeScript<'a> {
+            }
+            Expression::StringLiteral(lit) => Some(ConstantValue::String(lit.value.to_string())),
+            Expression::TemplateLiteral(lit) => {
+                let mut value = String::new();
+                for part in &lit.quasis {
+                    value.push_str(&part.value.raw);
+                }
+                Some(ConstantValue::String(value))
+            }
+            _ => None,
+        }
+    }
+
+    #[allow(clippy::cast_possible_truncation, clippy::cast_precision_loss)]
+    fn eval_binary_expression(
+        &self,
+        expr: &BinaryExpression<'a>,
+        prev_members: &FxHashMap<Atom<'a>, ConstantValue>,
+    ) -> Option<ConstantValue> {
+        let left = self.evaluate(&expr.left, prev_members)?;
+        let right = self.evaluate(&expr.right, prev_members)?;
+
+        if matches!(expr.operator, BinaryOperator::Addition)
+            && (matches!(left, ConstantValue::String(_))
+                || matches!(right, ConstantValue::String(_)))
+        {
+            let left_string = match left {
+                ConstantValue::NumericLiteral((raw, _)) => raw.to_string(),
+                ConstantValue::String(str) => str,
+                ConstantValue::F64(v) => v.to_string(),
+                ConstantValue::Identifier(_) => unreachable!(),
+            };
+
+            let right_string = match right {
+                ConstantValue::NumericLiteral((raw, _)) => raw.to_string(),
+                ConstantValue::String(str) => str,
+                ConstantValue::F64(v) => v.to_string(),
+                ConstantValue::Identifier(_) => unreachable!(),
+            };
+
+            return Some(ConstantValue::String(format!("{left_string}{right_string}")));
+        }
+
+        let left = match left {
+            ConstantValue::NumericLiteral((_, v)) | ConstantValue::F64(v) => v,
+            ConstantValue::String(_) | ConstantValue::Identifier(_) => return None,
+        };
+
+        let right = match right {
+            ConstantValue::NumericLiteral((_, v)) | ConstantValue::F64(v) => v,
+            ConstantValue::String(_) | ConstantValue::Identifier(_) => return None,
+        };
+
+        match expr.operator {
+            BinaryOperator::BitwiseOR => {
+                Some(ConstantValue::F64((left as i64 | right as i64) as f64))
+            }
+            BinaryOperator::BitwiseAnd => {
+                Some(ConstantValue::F64((left as i64 & right as i64) as f64))
+            }
+            BinaryOperator::ShiftRight | BinaryOperator::ShiftRightZeroFill => {
+                Some(ConstantValue::F64((left as i64 >> right as i64) as f64))
+            }
+            BinaryOperator::ShiftLeft => {
+                Some(ConstantValue::F64(((left as i64) << right as i64) as f64))
+            }
+            BinaryOperator::BitwiseXOR => {
+                Some(ConstantValue::F64((left as i64 ^ right as i64) as f64))
+            }
```

Comment:
I think we should use `i32` instead of `i64` here, considering the bitwise operations in JavaScript.

Reply:
Thank you, you're right! Now I know why I'm casting to `i32`, it's defined in the [spec](https://tc39.es/ecma262/#sec-numberbitwiseop).